Cannot use a metadataFilter except MetadataFilterChain (UnsupportedOperationException)

Defining a Metadata Filter, other than org.opensaml.saml2.metadata.provider.MetadataFilterChain to be used fails with UnsupportedOperationException upon application initialization:
```
Caused by: java.lang.UnsupportedOperationException
	at java.util.AbstractList.add(AbstractList.java:148) ~[?:1.8.0_45]
	at java.util.AbstractList.add(AbstractList.java:108) ~[?:1.8.0_45]
	at org.springframework.security.saml.metadata.MetadataManager.initializeProviderFilters(MetadataManager.java:525) 
...
```
This happens because the code in org.springframework.security.saml.metadata.MetadataManager assumes that the filters list from opensaml always allows addition:
```
if (currentFilter instanceof MetadataFilterChain) {
    log.debug("Adding signature filter into existing chain");
    MetadataFilterChain chain = (MetadataFilterChain) currentFilter;
    chain.getFilters().add(filter);
} else {
    log.debug("Combining signature filter with the existing in a new chain");
    MetadataFilterChain chain = new MetadataFilterChain();
    chain.getFilters().add(currentFilter);
    chain.getFilters().add(filter);
}
```
but this is not true on the else branch with a freshly created MetadataFilterChain. Even if allowed initially in opensaml the implementation of org.opensaml.saml2.metadata.provider.MetadataFilterChain is no longer initializing the filter list with an ArrayList, but rather with Collections.emptyList();

The workaround is to always use a wrapper MetadataFilterChain, and that's fine, however discovering the workaround takes a lot of time and fixing this hopefully saves other's time.

We detected this by trying to filter out Metada for SPs from, let's say InCommon, so we defined at top level an "idpOnly" bean as below:
```
<bean id="idpOnly" class="org.opensaml.saml2.metadata.provider.EntityRoleFilter">
	<constructor-arg>
  	<list value-type="javax.xml.namespace.QName">
  		<bean class="javax.xml.namespace.QName">
				 <constructor-arg value="urn:oasis:names:tc:SAML:2.0:metadata"/>
				 <constructor-arg value="IDPSSODescriptor"/>
  		</bean>
  	</list>
  </constructor-arg>
</bean>
```
and then referenced it in the "metadata" bean as below:
```
<bean class="org.springframework.security.saml.metadata.ExtendedMetadataDelegate">
	<constructor-arg>
	  <bean
	      class="org.opensaml.saml2.metadata.provider.ResourceBackedMetadataProvider">
	      <constructor-arg>
	          <bean class="java.util.Timer"/>
	      </constructor-arg>
	      <constructor-arg>
	          <bean class="org.opensaml.util.resource.HttpResource">
	              <constructor-arg value="http://md.incommon.org/InCommon/InCommon-metadata.xml"/>
	          </bean>
	      </constructor-arg>
	      <property name="parserPool" ref="parserPool"/>
	      <property name="metadataFilter" ref="idpOnly"/>                   
	  </bean>
	</constructor-arg>
	 <property name="metadataTrustCheck" value="false"/>
</bean>
```
Thank you and kind regards

--- a/core/src/main/java/org/springframework/security/saml/metadata/MetadataManager.java
+++ b/core/src/main/java/org/springframework/security/saml/metadata/MetadataManager.java
@@ -15,15 +15,7 @@
 package org.springframework.security.saml.metadata;
 
 import java.security.cert.X509Certificate;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Set;
-import java.util.Timer;
-import java.util.TimerTask;
+import java.util.*;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import org.opensaml.common.xml.SAMLConstants;
@@ -551,19 +543,12 @@ public class MetadataManager extends ChainingMetadataProvider implements Extende
 
             log.debug("Created new trust manager for metadata provider {}", provider);
 
-            // Combine any existing filters with the signature verification
             MetadataFilter currentFilter = provider.getMetadataFilter();
             if (currentFilter != null) {
-                if (currentFilter instanceof MetadataFilterChain) {
-                    log.debug("Adding signature filter into existing chain");
-                    MetadataFilterChain chain = (MetadataFilterChain) currentFilter;
-                    chain.getFilters().add(filter);
-                } else {
-                    log.debug("Combining signature filter with the existing in a new chain");
-                    MetadataFilterChain chain = new MetadataFilterChain();
-                    chain.getFilters().add(currentFilter);
-                    chain.getFilters().add(filter);
-                }
+                log.debug("Adding signature filter before existing filters");
+                MetadataFilterChain chain = new MetadataFilterChain();
+                chain.setFilters(Arrays.asList(filter, currentFilter));
+                provider.setMetadataFilter(chain);
             } else {
                 log.debug("Adding signature filter");
                 provider.setMetadataFilter(filter);

